chapters/grammar.adoc

```diff
@@ -9,84 +9,242 @@ The tokens returned from lexical analysis are
+| USAMPLERCUBEARRAY | `usamplerCubeArray`
+| SAMPLER2DMS | `sampler2DMS`
+| ISAMPLER2DMS | `isampler2DMS`
+| USAMPLER2DMS | `usampler2DMS`
+| SAMPLER2DMSARRAY | `sampler2DMSArray`
+| ISAMPLER2DMSARRAY | `isampler2DMSArray`
+| USAMPLER2DMSARRAY | `usampler2DMSArray`
+| IMAGE2D | `image2D`
+| IIMAGE2D | `iimage2D`
+| UIMAGE2D | `uimage2D`
+| IMAGE3D | `image3D`
+| IIMAGE3D | `iimage3D`
+| UIMAGE3D | `uimage3D`
+| IMAGECUBE | `imageCube`
+| IIMAGECUBE | `iimageCube`
+| UIMAGECUBE | `uimageCube`
+| IMAGEBUFFER | `imageBuffer`
+| IIMAGEBUFFER | `iimageBuffer`
+| UIMAGEBUFFER | `uimageBuffer`
+| IMAGE2DARRAY | `image2DArray`
+| IIMAGE2DARRAY | `iimage2DArray`
+| UIMAGE2DARRAY | `uimage2DArray`
+| IMAGECUBEARRAY | `imageCubeArray`
+| IIMAGECUBEARRAY | `iimageCubeArray`
+| UIMAGECUBEARRAY | `uimageCubeArray`
+| |
 ifdef::GLSL[]
-IMAGE1D IIMAGE1D UIMAGE1D
-IMAGE1DARRAY IIMAGE1DARRAY UIMAGE1DARRAY
-IMAGE2DRECT IIMAGE2DRECT UIMAGE2DRECT
-IMAGE2DMS IIMAGE2DMS UIMAGE2DMS
-IMAGE2DMSARRAY IIMAGE2DMSARRAY UIMAGE2DMSARRAY
+| IMAGE1D | `image1D`
+| IIMAGE1D | `iimage1D`
+| UIMAGE1D | `uimage1D`
+| IMAGE1DARRAY | `image1DArray`
+| IIMAGE1DARRAY | `iimage1DArray`
+| UIMAGE1DARRAY | `uimage1DArray`
+| IMAGE2DRECT | `image2DRect`
+| IIMAGE2DRECT | `iimage2DRect`
+| UIMAGE2DRECT | `uimage2DRect`
+| IMAGE2DMS | `image2DMS`
+| IIMAGE2DMS | `iimage2DMS`
+| UIMAGE2DMS | `uimage2DMS`
+| IMAGE2DMSARRAY | `image2DMSArray`
+| IIMAGE2DMSARRAY | `iimage2DMSArray`
+| UIMAGE2DMSARRAY | `uimage2DMSArray`
 endif::GLSL[]
-
-STRUCT VOID
-
-WHILE BREAK CONTINUE DO ELSE FOR IF DISCARD RETURN SWITCH CASE DEFAULT
-ifdef::GLSL[SUBROUTINE]
-
-IDENTIFIER TYPE_NAME
-FLOATCONSTANT INTCONSTANT UINTCONSTANT BOOLCONSTANT
-ifdef::GLSL[DOUBLECONSTANT]
-FIELD_SELECTION
-
-LEFT_OP RIGHT_OP
-INC_OP DEC_OP LE_OP GE_OP EQ_OP NE_OP
-AND_OP OR_OP XOR_OP MUL_ASSIGN DIV_ASSIGN ADD_ASSIGN
-MOD_ASSIGN LEFT_ASSIGN RIGHT_ASSIGN AND_ASSIGN XOR_ASSIGN OR_ASSIGN
-SUB_ASSIGN
-LEFT_PAREN RIGHT_PAREN LEFT_BRACKET RIGHT_BRACKET LEFT_BRACE RIGHT_BRACE DOT
-COMMA COLON EQUAL SEMICOLON BANG DASH TILDE PLUS STAR SLASH PERCENT
-LEFT_ANGLE RIGHT_ANGLE VERTICAL_BAR CARET AMPERSAND QUESTION
-
-INVARIANT PRECISE
-HIGH_PRECISION MEDIUM_PRECISION LOW_PRECISION PRECISION
+| |
+| STRUCT | `struct`
+| VOID | `void`
+| |
+| WHILE | `while`
+| BREAK | `break`
+| CONTINUE | `continue`
+| DO | `do`
+| ELSE | `else`
+| FOR | `for`
+| IF | `if`
+| DISCARD | `discard`
+| RETURN | `return`
+| SWITCH | `switch`
+| CASE | `case`
+| DEFAULT | `default`
+ifdef::GLSL[| SUBROUTINE | `subroutine` ]
+| |
+| IDENTIFIER | <<identifiers>>
+| TYPE_NAME | <<identifiers>>
+| FLOATCONSTANT | <<floats>>
+| INTCONSTANT | <<integers>>
+| UINTCONSTANT | <<integers>>
+| BOOLCONSTANT | <<booleans>>
+ifdef::GLSL[| DOUBLECONSTANT | <<floats>> ]
+| FIELD_SELECTION | <<identifiers>>
+| |
+| LEFT_OP | `<<`
+| RIGHT_OP | `>>`
+| INC_OP | `++`
+| DEC_OP | `--`
+| LE_OP | `\<=`
+| GE_OP | `>=`
+| EQ_OP | `==`
+| NE_OP | `!=`
+| AND_OP | `&`
```

Comment:
I just realized, these three are, possibly, wrong. I think they should actually be `&&`, `^^` and `||` ?

Reply:
Yes, these ought to be the logical version, not the bitwise one.

Reply:
Done.